### PR TITLE
Document our culture in more detail.

### PR DESCRIPTION
**Pull request by Hixie**
Partly aspirational.

### CODE REVIEW

File: design-principles.md [hunk 1]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+then we hope you will also follow these philosophies. Please also read
+our
+[style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo)
+for more specific guidelines regarding writing Dart code for Flutter.
+
+
+Code review and checking in code
+--------------------------------
+
+Every PR must be code-reviewed before check-in, including thing like
```
Comment (HansMuller) [inline design-principles.md]: thing -> things

**Comment by tvolkert** [comment]
LGTM

File: design-principles.md [hunk 2]
```diff
@@ -31,7 +82,11 @@ When you do implement a feature, implement it the right way. Avoid
 workarounds. Workarounds merely kick the problem further down the
 road, but at a higher cost: someone will have to relearn the problem,
 figure out the workaround and how to dismantle it (and all the places
-that now use it), _and_ implement the feature.
+that now use it), _and_ implement the feature. **Embrace the yak
+shave.** It's much better to take longer to fix a problem properly,
```
Comment (Hixie) [inline design-principles.md]: Yeah, I'll try to clarify.

File: design-principles.md [hunk 3]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+"LGTM" on your PR, and you have addressed all their feedback. ("LGTM"
+means "Looks Good To Me".) If you review a patch, you are sharing the
+responsibility for the patch with its author.
+
+Reviewers should not give an LGTM unless the patch has tests that
+verify all the affected code, or unless a test would make no sense.
+
+A reviewer may in some circumstances consider the code satisfactory
+without having fully reviewed or understood it. If a reviewer has not
+fully reviewed the code, they must admit to this by saying "RSLGTM"
```
Comment (eseidelGoogle) [inline design-principles.md]: should admit? :)  Unless you're trying for spec language.

File: design-principles.md [hunk 4]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+rather than just "LGTM". If you feel your code needs a real review,
+please find someone to actually review it. ("RSLGTM" means "Rubber
+Stamp Looks Good To Me".) If you mark a patch as RSLGTM, you are still
```
Comment (eseidelGoogle) [inline design-principles.md]: any punctuation between rubber stamp and lgtm?

File: design-principles.md [hunk 5]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+sharing the responsibility for the patch with its author.
+
+If you really need to check something in in a hurry, for example
+because everything is broken and you can fix it, then pick someone on
+the team who you want to have review the code, and then mark the PR as
+"TBR" with their name. ("TBR" means "To Be Reviewed".) This is only to
+be used in emergencies. Nobody being around to review your 50,000 line
```
Comment (eseidelGoogle) [inline design-principles.md]: I might make "nobody..." a parenthetical, it sounds pretty cynical.  So far I'd like to assume that all our community have good intentions. :)

File: design-principles.md [hunk 6]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+patch at midnight on December 31st is not an emergency.
+
+Run the tests before checking code in. Travis does this for you, so
```
Comment (eseidelGoogle) [inline design-principles.md]: The first sentence is confusing, I'd start by saying waiting for Travis instead of asking for folks to do something.

File: design-principles.md [hunk 7]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+wait for Travis to give the green light before merging a PR. The only
+time it is ok to skip this step is if the tests are already failing,
+and your fix is intended to improve matters, and you're going to watch
+the results once you've checked it in (in which case the bots are
+essentially doing the same job as Travis would have on your PR).
+
+Make sure all the trees and dashboards are green before checking in:
```
Comment (eseidelGoogle) [inline design-principles.md]: Really?  I'm supposed to check 5 sites before committing?  I'm glad we're writing it down, but seems like we should fix this.

File: design-principles.md [hunk 8]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+the [infra waterfall](https://build.chromium.org/p/client.flutter/waterfall),
+our [travis dashboard](https://travis-ci.org/flutter/flutter/builds),
+our [test dashboard](https://flutter-dashboard.appspot.com/build.html) (Google-only, sorry),
+our [status dashboard](https://flutter-dashboard.appspot.com/status.html) (Google-only, sorry), and
+our [benchmarks dashboard](https://flutter-dashboard.appspot.com/benchmarks.html) (Google-only, sorry).
+
+**If the trees or dashboards are showing any regressions, only fixes
+that improve the situation are allowed to go in**.
```
Comment (eseidelGoogle) [inline design-principles.md]: Do we enforce this?

File: design-principles.md [hunk 9]
```diff
@@ -8,14 +8,65 @@ permalink: /design-principles/
...
+then we hope you will also follow these philosophies. Please also read
+our
+[style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo)
+for more specific guidelines regarding writing Dart code for Flutter.
+
+
+Code review and checking in code
+--------------------------------
+
+Every PR must be code-reviewed before check-in, including things like
```
Comment (eseidelGoogle) [inline design-principles.md]: I might talk a bit about the purpose of code review, or link to something that does.  For me, a big part of it is sharing knowledge.  That's why even the most experienced engineers get their patches reviewed, so that through the process they are sharing the knowledge of their patch with someone else.  Then always 2 people understand some system instead of one.  Obviously there is also the correctness validation, but to me that's secondary to the bi-directional information sharing/teaching.

File: design-principles.md [hunk 10]
```diff
@@ -31,7 +82,14 @@ When you do implement a feature, implement it the right way. Avoid
 workarounds. Workarounds merely kick the problem further down the
 road, but at a higher cost: someone will have to relearn the problem,
 figure out the workaround and how to dismantle it (and all the places
-that now use it), _and_ implement the feature.
+that now use it), _and_ implement the feature. It's much better to
+take longer to fix a problem properly, than to be the one who fixes
+everything quickly but in a way that will require cleaning up later.
+
+You may hear team members say "embrace the yak shave!". This is an
```
Comment (eseidelGoogle) [inline design-principles.md]: A non-english speaker might find this confusing.  Do we have something to link to here?

File: design-principles.md [hunk 11]
```diff
@@ -41,8 +99,21 @@ and verify the test passes.
...
 When you implement a new feature, write tests for it.
 
-Run the tests before checking code in. (Travis does this for you, so
-wait for Travis to give the green light before merging a PR.)
+
+Documentation
+-------------
+
+If you have a question about our systems, then when you find the
+answer, put it into the documentation in the same place where you
```
Comment (eseidelGoogle) [inline design-principles.md]: please write a PR to add it.... :)  The "please" is the part I most wanted to add here.

File: design-principles.md [hunk 12]
```diff
@@ -92,10 +163,18 @@ API design
...
+Only assign a bug to yourself when you are actively working on it. If
+you're not working on it, leave it unassigned. Don't assign bugs to
+people unless you know they are going to work on it.
+
+You may hear team members refer to "licking the cookie". Assigning a
+bug to yourself, or otherwise indicating that you will work on it,
+tells others on the team to not fix it. If you then don't work on it
+straight away, you are acting like someone who has taken a cookie,
+licked it to unappetising to other people, and then not eaten it right
+away. By extension, "unlicking the cookie" means indicating to the
```
Comment (eseidelGoogle) [inline design-principles.md]: I've never heard of unlicking the cookie. :)  But it's not a bad idea to mention the idea that it's good to unassign bugs you aren't working on so others feel more empowered to take them?

File: design-principles.md [hunk 13]
```diff
@@ -109,6 +188,12 @@ check-in (even if it isn't yours) unless doing so would take longer
 than fixing the bug. When the trunk is broken, it slows down everyone
 else on the project.
 
+If things are broken, your priority should be helping the team fix the
```
Comment (eseidelGoogle) [inline design-principles.md]: "our" instead of "your"?  Less prescriptive, more embracing :)

File: design-principles.md (hunk 9 above)
Comment (Hixie) [inline design-principles.md]: good call

File: design-principles.md (hunk 4 above)
Comment (Hixie) [inline design-principles.md]: maybe hyphens, as in "Rubber stamp looks-good-to-me"? :-)

File: design-principles.md (hunk 5 above)
Comment (Hixie) [inline design-principles.md]: Agreed.

File: design-principles.md (hunk 6 above)
Comment (Hixie) [inline design-principles.md]: agreed

File: design-principles.md (hunk 8 above)
Comment (Hixie) [inline design-principles.md]: This document is the enforcement.

File: design-principles.md (hunk 10 above)
Comment (Hixie) [inline design-principles.md]: i'll link to a definition, but really this whole paragraph is just trying to explain the term, so...

File: design-principles.md (hunk 11 above)
Comment (Hixie) [inline design-principles.md]: This is aimed at team members, not our users. I'll rephrase it to make that clear.

File: design-principles.md (hunk 12 above)
Comment (Hixie) [inline design-principles.md]: will clarify

File: design-principles.md (hunk 13 above)
Comment (Hixie) [inline design-principles.md]: will rephrase
